File: dsa/li10_tree/bt/java/Bt_L199_BinaryTreeRightSideView.java

import java.util.ArrayList;
import java.util.List;

import com.algos.topics.models.TreeNode;


public class Bt_L199_BinaryTreeRightSideView {

	public static void main(String[] args) {
		Bt_L199_BinaryTreeRightSideView obj = new Bt_L199_BinaryTreeRightSideView();

        TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

		System.out.println(obj.rightSideView(root));
	}

    // Problem: https://leetcode.com/problems/binary-tree-right-side-view/
    // Idea: Apply inorder with right first. Pass level as arg to function.
    // At each level if size of result is equal to level add that element. Once an element already added at right side, there won't be any element gets added left side as result list size would have increased.
    // Time: O(n)
    // space: O(1)
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<Integer>();
        preorder(root, 0, ans);
        return ans;
    }

    private void preorder(TreeNode root, int level, List<Integer> ans){
        if(root == null)
            return;

        if(ans.size() == level)
            ans.add(root.val);

        preorder(root.right, level+1, ans);
        preorder(root.left, level+1, ans);
    }
}

File: dsa/patterns/10_trees/bt/python/bt_l199__binary_tree_right_side_view.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/binary-tree-right-side-view/
# Idea: Preorder traversal with right child first. Pass level as argument.
# At each level, if size of result equals level, add element.
# Once an element is added at right side, left-side elements are ignored for that level.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def right_side_view(root):
    ans = []

    def preorder(node, level):
        if not node:
            return

        if len(ans) == level:
            ans.append(node.val)

        preorder(node.right, level + 1)
        preorder(node.left, level + 1)

    preorder(root, 0)
    return ans


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(right_side_view(root))  # Expected: [1, 4, 6, 5]
